File: GOL-Java/BarszczSosnowskiego.java
package pl.edu.pg.eti.ksg.po.project2.rosliny;

import pl.edu.pg.eti.ksg.po.project2.Organizm;
import pl.edu.pg.eti.ksg.po.project2.Roslina;
import pl.edu.pg.eti.ksg.po.project2.Komentator;
import pl.edu.pg.eti.ksg.po.project2.Swiat;
import pl.edu.pg.eti.ksg.po.project2.Punkt;

import java.awt.*;
import java.util.Random;

public class BarszczSosnowskiego extends Roslina {
    private static final int sila = 10;
    private static final int inicjatywa = 0;

    public BarszczSosnowskiego(Swiat swiat, Punkt pozycja, int turaUrodzenia) {
        super(TypOrganizmu.BARSZCZ_SOSNOWSKIEGO, swiat, pozycja, turaUrodzenia, sila, inicjatywa);
        setKolor(Color.BLACK);
        setSzansaRozmnazania(0.05);
    }

    @Override
    public void Akcja() {
        int pozX = getPozycja().getX();
        int pozY = getPozycja().getY();
        LosujPoleDowolne(getPozycja());// BLOKUJE GRANICE
        for (int i = 0; i < 4; i++) {
            Organizm tmpOrganizm = null;
            if (i == 0 && !CzyKierunekZablokowany(Kierunek.DOL))
                tmpOrganizm = getSwiat().CoZnajdujeSieNaPolu(new Punkt(pozX, pozY + 1));
            else if (i == 1 && !CzyKierunekZablokowany(Kierunek.GORA))
                tmpOrganizm = getSwiat().CoZnajdujeSieNaPolu(new Punkt(pozX, pozY - 1));
            else if (i == 2 && !CzyKierunekZablokowany(Kierunek.LEWO))
                tmpOrganizm = getSwiat().CoZnajdujeSieNaPolu(new Punkt(pozX - 1, pozY));
            else if (i == 3 && !CzyKierunekZablokowany(Kierunek.PRAWO))
                tmpOrganizm = getSwiat().CoZnajdujeSieNaPolu(new Punkt(pozX + 1, pozY));

            if (tmpOrganizm != null && tmpOrganizm.CzyJestZwierzeciem()) {
                getSwiat().UsunOrganizm(tmpOrganizm);
                Komentator.DodajKomentarz(OrganizmToString() + " zabija " + tmpOrganizm.OrganizmToString());
            }
        }
        Random rand = new Random();
        int tmpLosowanie = rand.nextInt(100);
        if (tmpLosowanie < getSzansaRozmnazania() * 100) Rozprzestrzenianie();
    }

    @Override
    public String GetGatunek() {
        return "Barszcz Sosnowskiego";
    }

    @Override
    public boolean SpecjalneDzialaniePodczasAtaku(Organizm atakujacy, Organizm ofiara) {
        if (atakujacy.getSila() >= 10) {
            getSwiat().UsunOrganizm(this);
            Komentator.DodajKomentarz(atakujacy.OrganizmToString() + " zjada " + this.OrganizmToString());
            atakujacy.WykonajRuch(ofiara.getPozycja());
        }
        return true;
    }
}

File: GOL-C++/matrix/BarszczSosnowskiego.cpp
#include <iostream>
#include <typeinfo>

#include "Swiat.h"
#include "Zwierze.h"
#include "Roslina.h"
#include "BarszczSosnowskiego.h"

BarszczSosnowskiego::BarszczSosnowskiego() : Roslina() {
	sila = 10;
}

BarszczSosnowskiego::BarszczSosnowskiego(int x, int y, Swiat* swiat) : Roslina(10, x, y, swiat) {}

char BarszczSosnowskiego::GetAwatar() const {
	return awatar;
}

string BarszczSosnowskiego::GetGatunek() const {
	return "Barszcz Sosnowskiego";
}

void BarszczSosnowskiego::akcja() {
	Organizm* org[4];
	org[0] = swiat->GetLokalizacja(x, y - 1);
	org[1] = swiat->GetLokalizacja(x + 1, y);
	org[2] = swiat->GetLokalizacja(x, y + 1);
	org[3] = swiat->GetLokalizacja(x - 1, y);

	if ((org[0] != nullptr) && org[0]->CzyZwierze())
		swiat->zniszczOrganizm(org[0]);
	if ((org[1] != nullptr) && org[1]->CzyZwierze())
		swiat->zniszczOrganizm(org[1]);
	if ((org[2] != nullptr) && org[2]->CzyZwierze())
		swiat->zniszczOrganizm(org[2]);
	if ((org[3] != nullptr) && org[3]->CzyZwierze())
		swiat->zniszczOrganizm(org[3]);
	int ran = (rand() % 100) + 1;
	if (ran < ratio)
		Roslina::akcja();
}

void BarszczSosnowskiego::kolizja(Organizm* atakujacy) {
	swiat->zniszczOrganizm(atakujacy);
	swiat->zniszczOrganizm(this);
}